lib/spack/spack/schema/env.py

```diff
@@ -34,22 +67,53 @@
                             'type': 'string'
                         },
                     },
-                    'specs': {
-                        # Specs is a list of specs, which can have
-                        # optional additional properties in a sub-dict
+                    'definitions': {
                         'type': 'array',
                         'default': [],
-                        'additionalProperties': False,
                         'items': {
-                            'anyOf': [
-                                {'type': 'string'},
-                                {'type': 'null'},
-                                {'type': 'object'},
-                            ]
+                            'type': 'object',
+                            'properties': {
+                                'when': {
+                                    'type': 'string'
+                                }
+                            },
+                            'patternProperties': {
+                                '^(?!when$)\w*': spec_list_schema
```

Comment:
This should be a raw string as it contains a regex:
```suggestion
                               r'^(?!when$)\w*': spec_list_schema
```